crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1262,6 +1262,23 @@ impl<'db> Type<'db> {
                 Type::Callable(CallableType::WrapperDescriptorDunderGet),
             ) => !KnownClass::WrapperDescriptorType.is_subclass_of(db, class),
 
+            (
+                Type::Callable(CallableType::General(_)) | Type::FunctionLiteral(_),
+                Type::Callable(CallableType::General(_)),
+            )
+            | (Type::Callable(CallableType::General(_)), Type::FunctionLiteral(_)) => {
```

Comment:
What's the reason for the asymmetry here? Could this be

```suggestion
            (
                Type::Callable(CallableType::General(_)) | Type::FunctionLiteral(_),
                Type::Callable(CallableType::General(_)) | Type::FunctionLiteral(_),
            ) => {
```

or

```suggestion
            (
                Type::Callable(CallableType::General(_)) | Type::FunctionLiteral(_),
                Type::Callable(CallableType::General(_)),
            )
            | (
                Type::Callable(CallableType::General(_)),
                Type::Callable(CallableType::General(_)) | Type::FunctionLiteral(_),
            ) => {
```

Reply:
The second suggestion could work, but not the first. Two `FunctionLiteral` types can definitely be disjoint from each other (in fact they always are.)

I think the reason for the asymmetry in the current version is simply avoiding redundancy. The case of two general-callable-types is always handled by the first clause, so doesn't need to be represented in the second clause.

I don't think it matters much but IMO the clearest version might be to have three separate clauses: function-literal vs general-callable, general-callable vs function-literal, and general-callable vs general-callable.

Reply:
> The second suggestion could work, but not the first. Two `FunctionLiteral` types can definitely be disjoint from each other (in fact they always are.)

Yes, and they're handled a bit above in the match expression.

> I think the reason for the asymmetry in the current version is simply avoiding redundancy. The case of two general-callable-types is always handled by the first clause, so doesn't need to be represented in the second clause.

Yup, this is the reason (I was about to comment this :)).

> I don't think it matters much but IMO the clearest version might be to have three separate clauses: function-literal vs general-callable, general-callable vs function-literal, and general-callable vs general-callable.

I started out with this but clippy complains about "unnested or-patterns" which then suggests to collapse it into what the current pattern is in the PR. We would need to have separate branches for it to work but I think I'd want to avoid that so that the comment explanation remains present in the same branch for all related cases.